README.md
## Carousel component for react-native

### Installation
```bash
npm install react-native-carousel
```

### Properties

```js
hideIndicators={false} // Set to true to hide the indicators
indicatorColor="#FFFFFF" // Active indicator color
indicatorSize={20} // Indicator bullet size
indicatorSpace={15} // space between each indicator
inactiveIndicatorColor="#999999" // Inactive indicator color
indicatorAtBottom={true} // Set to false to show the indicators at the top
indicatorVerticalOffset={250} // Indicator relative position from top or bottom
indicatorHorizontalOffset={0} // Indicator relative position from center (horizontally)
onPageChange={callback} // Called when the active page changes
inactiveIndicatorText= '•' // Inactive indicator content ( You can customize to use any Unicode character )
indicatorText= '•' // Active indicator content ( You can customize to use any Unicode character )
animate={true} // Enable carousel autoplay
delay={1000} // Set Animation delay between slides
loop={true} // Allow infinite looped animation. Depends on Prop {...animate} set to true.
 
```

### Usage example

Assuming you have `npm install -g react-native-cli`, first generate an app:

    react-native init RNCarousel
    cd RNCarousel
    npm install react-native-carousel --save

Then paste the following into `RNCarousel/index.ios.js`:

```javascript
'use strict';

var React = require('react-native');
var {
  AppRegistry,
  StyleSheet,
  Text,
  View,
} = React;

var Carousel = require('react-native-carousel');

var RNCarousel = React.createClass({
  render: function() {
    return (
      <Carousel width={375}>
        <View style={styles.container}>
          <Text>Page 1</Text>
        </View>
        <View style={styles.container}>
          <Text>Page 2</Text>
        </View>
        <View style={styles.container}>
          <Text>Page 3</Text>
        </View>
      </Carousel>
    );
  }
});

var styles = StyleSheet.create({
  container: {
    width: 375,
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'transparent',
  },
});

AppRegistry.registerComponent('RNCarousel', () => RNCarousel);
```


CHANGELOG.md
### 0.11.0
- React 16 support (@abuinitski)

### 0.10.0
- Fix horizontal scroll in RN v0.28 (@sghiassy, @dmccer)

### 0.9.1
- onPageChange never called on indicator press on Android (@lideen)

### 0.9.0
- Fix carousel in iOS with react-native 0.28 (@francisco-sanchez-molina)

### 0.8.0
- Custom indicator content (@shidhincr)

### 0.7.0
- Call createClass on React instead of React Native (@scottluptowski)

### 0.6.1
- Hide inactive indicators (@scottluptowski)

### 0.6.0
- Compute width on the fly (@mikelambert)
- Android support (@mikelambert, @kushal)

### 0.5.0
- Fix for pressing indicator with custom width (@juankiz)
- Fix scrollWithoutAnimationTo deprecation warning (@juankiz)

### 0.4.6
- Fix depreciated parameter for scrollTo (@ronaldx)

### 0.4.5
- Fix scrolling to top when pressing iOS status bar (@juankiz)

### 0.4.4
- Added indicatorOffset property (@regeda)

### 0.4.3
- Fix for page loop animation

### 0.4.2
- Fix for Dimension import

### 0.4.1
- Remove duplicate width calculation when initialPage is defined (@ptmt)

### 0.4.0
- Add autoplay and looping options (@PraiseGeek)

### 0.3.5
- Add `indicatorSpace` prop (@alinz)

### 0.3.3
- Add `indicatorSize` prop (@daviferreira)

### 0.3.2
- Add `hideIndicators` prop (@daviferreira)

### 0.3.1
- Add `onPageChange` prop (@unknownexception)

### 0.3.0
- Bump react-native dependency

### 0.2.0
- Add `indicatorAtBottom` prop (@imaffett)
- Added new styles for the  indicator positions (@imaffett)
- Clicking the indicators will switch the card/page it is on. (@imaffett)
- Set the background of the indicator to transparent. (@imaffett)
- Set bounces=false on the ScrollView used for the Carousel (@imaffett)

### 0.1.0
- Initial component


index.js
'use strict';

var React = require('react-native');
var {
  StyleSheet,
  View,
  Text,
  ScrollView,
} = React;

var Carousel = React.createClass({

  getDefaultProps() {
    return {
      width: 375,
      indicatorColor: '#000000',
      inactiveIndicatorColor: '#999999'
    };
  },

  getInitialState() {
    return {
      activePage: 0
    };
  },

  render() {

    return (
      <View style={{ flex: 1 }}>
        <ScrollView
          contentContainerStyle={styles.container}
          automaticallyAdjustContentInsets={false}
          horizontal={true}
          pagingEnabled={true}
          showsHorizontalScrollIndicator={false}
          onMomentumScrollEnd={this.onAnimationEnd}
        >
          {this.props.children}
        </ScrollView>
        {this.renderPageIndicator()}
      </View>
    );
  },

  renderPageIndicator() {
    var indicators = [],
        style, position;

    position = {
      width: this.props.children.length * 15,
    };
    position.left = (this.props.width - position.width) / 2;

    for (var i=0; i< this.props.children.length; i++) {
      style = i === this.state.activePage ? { color: this.props.indicatorColor } : { color: this.props.inactiveIndicatorColor };
      indicators.push(<Text style={style}>&bull;</Text>);
    }

    return (
      <View style={[styles.pageIndicator, position]}>
        {indicators}
      </View>
    );
  },

  onAnimationEnd(e) {
    var activePage = e.nativeEvent.contentOffset.x / this.props.width;
    this.setState({
      activePage: activePage
    });
  }
});

var styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
  },
  page: {
    alignItems: 'center',
    justifyContent: 'center',
    borderWidth: 1,
  },
  pageIndicator: {
    position: 'absolute',
    flexDirection: 'row',
    flex: 1,
    bottom: 20,
    justifyContent: 'space-around',
    alignItems: 'center'
  },
});

module.exports = Carousel;


CarouselPager.ios.js
var React = require('react');
var createReactClass = require('create-react-class');
var {
  ScrollView,
} = require('react-native');

var CarouselPager = createReactClass({

  scrollToPage(page, animated) {
    if (typeof animated === 'undefined') {
      animated = true;
    }
    this.refs.scrollView.scrollTo({x: page * this.props.width, y: 0, animated: animated});
  },

  _onMomentumScrollEnd(e) {
    var activePage = e.nativeEvent.contentOffset.x / this.props.width;
    this.props.onEnd(activePage);
  },

  render() {
    return <ScrollView ref="scrollView"
      contentContainerStyle={this.props.contentContainerStyle}
      automaticallyAdjustContentInsets={false}
      horizontal={true}
      pagingEnabled={true}
      showsHorizontalScrollIndicator={false}
      bounces={false}
      onScrollBeginDrag={this.props.onBegin}
      onMomentumScrollEnd={this._onMomentumScrollEnd}
      scrollsToTop={false}
    >
      {this.props.children}
    </ScrollView>;
  },
});

module.exports = CarouselPager;


CarouselPager.android.js
var React = require('react');
var createReactClass = require('create-react-class');
var {
  View
} = require('react-native');
import ViewPager from "@react-native-community/viewpager";

var CarouselPager = createReactClass({

  scrollToPage(page, animated) {
    if (typeof animated === 'undefined') {
      animated = true;
    }
    if (animated) {
      this.refs.viewPager.setPage(page);
    } else {
      this.refs.viewPager.setPageWithoutAnimation(page);
    }
    this._onPageSelected(page);
  },

  _onPageSelected(page) {
    this.props.onEnd(page);
  },

  render() {
    return <ViewPager
      ref="viewPager"
      style={{flex: 1}}
      contentContainerStyle={this.props.contentContainerStyle}
      automaticallyAdjustContentInsets={false}
      horizontal={true}
      showsHorizontalScrollIndicator={false}
      bounces={false}
      onPageScroll={this.props.onBegin}
      onPageSelected={(e) => this._onPageSelected(e.nativeEvent.position)}
      scrollsToTop={false}
      >
        {this.props.children.map((c, idx) => <View key={idx} style={{flex: 1}}>{c}</View>)}
      </ViewPager>;
  },
});

module.exports = CarouselPager;


Carousel.js
'use strict';

var React = require('react');
var createReactClass = require('create-react-class');
var {
  Dimensions,
  StyleSheet,
  Text,
  View,
} = require('react-native');

var TimerMixin = require('react-timer-mixin');
var CarouselPager = require('./CarouselPager');

var Carousel = createReactClass({
  mixins: [TimerMixin],

  getDefaultProps() {
    return {
      hideIndicators: false,
      indicatorColor: '#000000',
      indicatorSize: 50,
      inactiveIndicatorColor: '#999999',
      indicatorAtBottom: true,
      indicatorPosition:20,
      indicatorOffset: 250,
      indicatorText: '•',
      inactiveIndicatorText: '•',
      width: null,
      initialPage: 0,
      indicatorSpace: 25,
      animate: true,
      delay: 1000,
      loop: true,
    };
  },

  getInitialState() {
    return {
      activePage: this.props.initialPage > 0 ? this.props.initialPage : 0,
    };
  },

  getWidth() {
    if (this.props.width !== null) {
      return this.props.width;
    } else {
      return Dimensions.get('window').width;
    }
  },

  componentDidMount() {
    if (this.props.initialPage > 0) {
      this.refs.pager.scrollToPage(this.props.initialPage, false);
    }

    if (this.props.animate && this.props.children){
        this._setUpTimer();
    }
  },

  indicatorPressed(activePage) {
    this.setState({activePage});
    this.refs.pager.scrollToPage(activePage);
  },

  renderPageIndicator() {
    if (this.props.hideIndicators === true) {
      return null;
    }
    
    var pageIndicatorStyles = StyleSheet.create({
        pageIndicatorTop: {
            top: this.props.indicatorPosition
        },
        pageIndicatorBottom: {
            bottom: this.props.indicatorPosition
        }
    });
        
    var indicators = [],
        indicatorStyle = this.props.indicatorAtBottom ? pageIndicatorStyles.pageIndicatorBottom : pageIndicatorStyles.pageIndicatorTop,
        style, position;

    position = {
      width: this.props.children.length * this.props.indicatorSpace,
    };
    position.left = (this.getWidth() - position.width) / 2;

    for (var i = 0, l = this.props.children.length; i < l; i++) {
      if (typeof this.props.children[i] === "undefined") {
        continue;
      }

      style = i === this.state.activePage ? { color: this.props.indicatorColor } : { color: this.props.inactiveIndicatorColor };
      indicators.push(
         <Text
            style={[style, { fontSize: this.props.indicatorSize }]}
            key={i}
            onPress={this.indicatorPressed.bind(this,i)}
          >
             { i === this.state.activePage  ? this.props.indicatorText : this.props.inactiveIndicatorText }
          </Text>
      );
    }

    if (indicators.length === 1) {
      return null;
    }

    return (
      <View style={[styles.pageIndicator, position, indicatorStyle]}>
        {indicators}
      </View>
    );
  },

  _setUpTimer() {
     if (this.props.children.length > 1) {
         this.clearTimeout(this.timer);
         this.timer = this.setTimeout(this._animateNextPage, this.props.delay);
     }
  },

  _animateNextPage() {
     var activePage = 0;
     if (this.state.activePage < this.props.children.length - 1) {
         activePage = this.state.activePage + 1;
     } else if (!this.props.loop) {
         return;
     }

     this.indicatorPressed(activePage);
     this._setUpTimer();
  },

  _onAnimationBegin() {
     this.clearTimeout(this.timer);
  },

  _onAnimationEnd(activePage) {
    this.setState({activePage});
    if (this.props.onPageChange) {
      this.props.onPageChange(activePage);
    }
  },

  render() {
    return (
      <View style={{ flex: 1 }}>
        <CarouselPager
          ref="pager"
          width={this.getWidth()}
          contentContainerStyle={styles.container}
          onBegin={this._onAnimationBeginPage}
          onEnd={this._onAnimationEnd}
        >
          {this.props.children}
        </CarouselPager>
        {this.renderPageIndicator()}
      </View>
    );
  },

});

var styles = StyleSheet.create({
  pageIndicator: {
    position: 'absolute',
    flexDirection: 'row',
    flex: 1,
    justifyContent: 'space-around',
    alignItems: 'center',
    alignSelf: 'center',
    backgroundColor:'transparent',
  },
});

module.exports = Carousel;
